store/postgres/src/store.rs

```diff
@@ -665,17 +704,116 @@ impl Store {
         Ok(())
     }
 
+    /// Panics:
+    /// * This task will panic if the supplied closure panics
+    /// * This task will panic if the supplied closure returns Err(Cancelled)
+    ///   when the supplied cancel token is not cancelled.
+    // Ideally, we want to pass a borrow of the PooledConnection in. But,
+    // with_entity_conn requires this to be owned to give it to e::Connection
+    // If e::Connection could take a reference this would not be a problem
+    async fn with_conn<T: Send + 'static>(
```

Comment:
This function and `with_entity_conn` is fundamental enough that it deserves documentation on how to use it properly, in particular, for which operations to use it and for which ones not to use it and instead fall back to `Store::get_conn`.

Your uses call `cancel_handle.check_cancel()?` as the first thing - is that necessary seeing how `with_conn` calls that right before calling `f`. As a user of this, when should I be calling  `cancel_handle.check_cancel()?` ?

Reply:
You're right. I'll add copious amounts of documentation around `with_conn` and `with_entity_conn` that encompasses a lot of the earlier discussions.

I don't think there is any case where one should fall back on `Store::get_conn`. The one place where it might make seem to make sense _right now_ is where you need multiple connections simultaneously (eg: a normal connection and an entity connection, or multiple entity connections) when you're already inside `with_conn`. But, there is a better way to deal with that for a slightly different API. Instead of using `with_entity_conn` as it exists today, simply have `with_conn` return some kind of wrapper type that has a method `to_entity_conn<'a>(&'a self) -> e::Connection<'a>` which would allow for re-use of the same connection multiple times in the same `with_conn` closure.

I wish I had time to port the rest of the database code and refine the API, but I think this is going to have to wait.

Reply:
Screw it. Maybe I'll go ahead and make that API since now having the good API doesn't take long, it's not inferring that everything needs to be migrated.

Reply:
Discussed offline, will not make these changes yet.